samples: drivers: CAN: reworked sample code

alexanderwachter opened the pull request:
The CAN sample is rewritten to be board independent.
The CAN controller and LED is determined automatically and the button is removed

karstenkoenig committed:
drivers: can: mcp2515: fix devicetree bindings

Adding required fields to the devicetree overlay of the CAN sample as
this is often used as a reference. Also use these fields instead of the
KConfig entries.

Signed-off-by: Karsten Koenig <karsten.koenig.030@gmail.com>

Commit:
samples: drivers: CAN: reworked sample code

The CAN sample is rewritten to be board independent.
The CAN controller and LED is determined automatically and the
button is removed.

Signed-off-by: Alexander Wachter <alexander.wachter@student.tugraz.at>

alexanderwachter commented:
@karstenkoenig ping. Can't add you as a reviewer :-(. Can you add yourself?

jukkar commented on samples/drivers/CAN/src/main.c:
```diff
@@ -208,73 +136,52 @@ void main(void)
-					 tx_thread,
+	printk("Change LED filter ID: %d\n", ret);
+
+	k_tid_t rx_tid = k_thread_create(&rx_thread_data, rx_thread_stack,
```
Variable definitions at the beginning of the functions.

alexanderwachter replied:
Thanks.

jukkar reviewed (commented):
LGTM, minor nit picking only.

karstenkoenig commented on samples/drivers/CAN/src/main.c:
```diff
@@ -208,73 +136,52 @@ void main(void)
+		UNALIGNED_PUT(sys_cpu_to_be16(counter),
+			      (u16_t *)&counter_frame.data[0]);
+		counter++;
+		/* This sending call is blocking until the message is sent. */
```
Technically this is blocking until timeout or sent, could probably just live without the comments, the intention throughout the sample is quite clear to cover two tx paths (blocking/non blocking) and two rx paths (work queue and polling on msgq) :-)

alexanderwachter replied:
@karstenkoenig the timeout is just for the mailbox acquisition. If the bus is busy, it waits "forever".

karstenkoenig replied:
Hmm right, yeah I remember I just followed the stm32 can driver there and try acquiring a tx buffer slot with a timeout. The blocking nature of the call comes from not providing a tx finished callback. 

Not sure about changing the behavior, right now there is no abort tx (except reinit the driver) and I'd just wait until the need/requirement comes up to have a use case to validate :-) I haven't seen a need yet.

karstenkoenig reviewed (approved):
Great changes!
Tested this on a nrf52_pca10040 + mcp2515 and it works with only configuring the mcp2515 overlay, no project config or further board config needed.

alexanderwachter commented:
@karstenkoenig removed the config for mcp2515.

alexanderwachter commented:
Ping @jfischer-phytec-iot. I added the ree_board CAN expansion as a shield.

Commit:
baords: shields: Add reel_board CANbus shield

This commit adds the reel_board canbus shield.
On the shield, there is an MCP2515 and a CAN transceiver.

Signed-off-by: Alexander Wachter <alexander.wachter@student.tugraz.at>

jfischer-phytec-iot commented on drivers/can/Kconfig.mcp2515:
```diff
@@ -20,34 +20,6 @@ config CAN_MCP2515_OSC_FREQ
-	default 2
-	range 1 8
-	help
-	  Time quanta of propagation segment (ISO 11898-1)
```
I think it is not correct to move this configs to device tree, does it depend on hardware design or is it software option which can be changed at given hardware?

alexanderwachter replied:
These are properties of the hardware. More specific, the transceiver, wiring, and other delays on the board.  It should have been in the device tree from the beginning.

karstenkoenig replied:
Welllll now they are 4 properties ;-) But yes, much nicer this way.
Also both KConfig and devicetree hardcode it, so it's not changing much in behavior

alexanderwachter commented:
`drivers: can: mcp2515: fix devicetree bindings` commit is from another PR. I just jerry picked it to get the CI passing.

dbkinder commented on samples/drivers/CAN/README.rst:
```diff
@@ -7,10 +7,9 @@ Overview
-Reception is indicated by blink LEDs and output to the console.
+Messages with standard and extended identifiers are sent over the bus.
+Messages are received using message-queues and work-queues.
+Reception is indicated by blink the LED (if present) and output to the console.
```
Change ``blink`` to ``blinking``

dbkinder commented on samples/drivers/CAN/README.rst:
```diff
@@ -19,25 +18,22 @@ In loopback mode, the board receives its own messages. This could be used for
 
 The sample can be built and executed on boards supporting CAN.
-The output ports and pins of the LEDs can be configured by Kconfig.
+The LED output pin is defined in boards device tree.
```
change to ``in the board's device tree.``

dbkinder commented on samples/drivers/CAN/README.rst:
```diff
@@ -19,25 +18,22 @@ In loopback mode, the board receives its own messages. This could be used for
-   Button pressed! Send message 0
-   Button pressed 4 times
-   String sent over CAN
+Change LED filter ID: 0
```
All the "console output" (lines 30-37) needs to be indented by three spaces

dbkinder commented on samples/drivers/CAN/README.rst:
```diff
@@ -19,25 +18,22 @@ In loopback mode, the board receives its own messages. This could be used for
-The output ports and pins of the LEDs can be configured by Kconfig.
+The LED output pin is defined in boards device tree.
+
+For the reel_board CAN expansion board, use the provided shield.
```
Can you provide a link reference to the "CAN expansion board" and "the provided shield".  I'm not sure what these are.  Who or what is providing the shield?

alexanderwachter replied:
Thanks for the review! The "CAN expansion board" is not yet available publicly.
I'll add the comment and links again when it's available.

henrikbrixandersen reviewed (approved):
LGTM. I have verified the revised sample on a setup using NXP TWR-KE18F boards, both in loopback and normal mode.

jfischer-phytec-iot commented on boards/shields/reel_board_canbus/reel_board_canbus.conf:
```diff
@@ -0,0 +1,3 @@
+CONFIG_SPI=y
+CONFIG_SPI_1=y
+CONFIG_CAN_MCP2515=y
```
neither CAN_1 nor CAN_0 is enabled, device label is "CAN_1", should CONFIG_CAN_1=y be here?

alexanderwachter replied:
Yes. Had to delete it in the proj.conf and forgot to add it here.

jfischer-phytec-iot commented on boards/shields/reel_board_canbus/reel_board_canbus.conf:
```diff
@@ -0,0 +1,3 @@
+CONFIG_SPI=y
```
Shield designation is "link board can", could you rename everything to `link_board_can`*, please?

alexanderwachter replied:
Sure

dbkinder reviewed (approved):
LGTM, but can you verify that ``uart:~$ `` is part of the output (line 32)

alexanderwachter commented:
@dbkinder yes, uart:~$ is there because of the shell.

alexanderwachter commented:
#14749 drivers/CAN can be ticked after merging :-)

ioannisg commented:
@alexanderwachter can you rebase to latest master?

alexanderwachter commented:
@ioannisg rebased.
